yrs/src/transact.rs

```diff
@@ -0,0 +1,235 @@
+                self.clone(),
+                store,
+                Some(origin.into()),
+            )),
+            None => Err(TransactionAcqError::ExclusiveAcqFailed),
+        }
+    }
+
+    fn transact_mut_with<T>(&self, origin: T) -> TransactionMut
+    where
+        T: Into<Origin>,
+    {
+        let lock = self.store.write_blocking();
+        TransactionMut::new(self.clone(), lock, Some(origin.into()))
+    }
+
+    fn transact(&self) -> Transaction {
+        let lock = self.store.read_blocking();
+        Transaction::new(lock)
+    }
+
+    fn transact_mut(&self) -> TransactionMut {
+        let lock = self.store.write_blocking();
+        TransactionMut::new(self.clone(), lock, None)
+    }
+}
+
+/// Trait implemented by [Doc] and shared types, used for carrying over the responsibilities of
+/// creating new transactions, used as a unit of work in Yrs.
+pub trait AsyncTransact<'doc> {
+    type Read: Future<Output = Transaction<'doc>>;
+    type Write: Future<Output = TransactionMut<'doc>>;
+
+    fn transact(&'doc self) -> Self::Read;
+    fn transact_mut(&'doc self) -> Self::Write;
+
+    /// Creates and returns a read-write capable transaction with an `origin` classifier attached.
+    /// This transaction can be used to mutate the contents of underlying document store and upon
+    /// dropping or committing it may subscription callbacks.
+    ///
+    /// An `origin` may be used to identify context of operations made (example updates performed
+    /// locally vs. incoming from remote replicas) and it's used i.e. by [`UndoManager`][crate::undo::UndoManager].
+    fn transact_mut_with<T>(&'doc self, origin: T) -> Self::Write
+    where
+        T: Into<Origin>;
+}
+
+impl<'doc> AsyncTransact<'doc> for Doc {
+    type Read = AcquireTransaction<'doc>;
+    type Write = AcquireTransactionMut<'doc>;
+
+    fn transact(&'doc self) -> Self::Read {
+        let fut = self.store.read_async();
+        AcquireTransaction { fut }
+    }
+
+    fn transact_mut(&'doc self) -> Self::Write {
+        let fut = self.store.write_async();
+        AcquireTransactionMut {
+            doc: self.clone(),
+            origin: None,
+            fut,
+        }
+    }
+
+    fn transact_mut_with<T>(&'doc self, origin: T) -> Self::Write
+    where
+        T: Into<Origin>,
+    {
+        let fut = self.store.write_async();
+        AcquireTransactionMut {
+            doc: self.clone(),
+            origin: Some(origin.into()),
+            fut,
+        }
+    }
+}
+
+pub struct AcquireTransaction<'doc> {
+    fut: Read<'doc, Store>,
+}
+
+impl<'doc> Unpin for AcquireTransaction<'doc> {}
+
+impl<'doc> Future for AcquireTransaction<'doc> {
+    type Output = Transaction<'doc>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let pinned = unsafe { Pin::new_unchecked(&mut self.fut) };
+        pinned.poll(cx).map(Transaction::new)
+    }
+}
+
+pub struct AcquireTransactionMut<'doc> {
```

Comment:
Do you think it's possible to export it so that we can `use yrs::AcquireTransactionMut`?

Reply:
Yes, I'll make an update.